File: Python/Binary Tree Zigzag Level Order Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def zigzagLevelOrder(self, root):
        if not root:
            return []
        level = 0
        ret = [[]]
        self.helper(root, level , ret)
        return ret
    
    def helper(self, root, level, ret):
        if not root:
            return None
        if level >= len(ret):
            ret.append([])
        if level % 2:
            ret[level].append(root.val)
        else:
            ret[level].insert(0, root.val)

        self.helper(root.right, level+1, ret)
        self.helper(root.left, level+1, ret)

        
        
        

File: Java/src/Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {

        int level = 0;
        
        List<List<Integer>> ret = new ArrayList<>();
        
        helper(root, level, ret);
        return ret;
    }
    
    private void helper(TreeNode root, int level, List<List<Integer>> ret){
        if (root == null) return;
        if (level >= ret.size()){
            List<Integer> newLevel = new ArrayList<>();
            ret.add(newLevel);
            
        } 
        if (level % 2 != 0){
            ret.get(level).add(root.val);
        } 
        else{
            ret.get(level).add(0, root.val);
        }
        helper(root.right, level+1, ret);
        helper(root.left, level+1, ret);
    }
}